README

- INFO
    - Tested under 10.04 LTS, 10.10 and 11.10 ubuntu w/python 2.6/2.7:
        1.2, 1.3, 1.4.1+

    - PLEASE NOTE: 
        * The latest master version will default to installing using github packages for swift
        * Read the ChangeLog file for more information
        * swift logging is now available


- INSTRUCTIONS 

    1. Obtain the code 
        - GITHUB TRUNK:
            git clone git://github.com/btorch/swift-saio.sh.git

        - GITHUB TAG:
            git clone git://github.com/btorch/swift-saio.sh.git TAG_VALUE

        - Or download a tagged version as a compressed archive from github

    2. After clonning the github repo or unconpressing the archive 
        - cd into the swift-saio.sh directory 

    3. Go over the swift-saio.cfg file and make any needed changes
        - Note that some Variables MUST be changed 

    4. Once you have edited the swift-saio.cfg file to your needs, proceed 
        - Now execute the main script from within the "swift-saio.sh" directory as the root user

        For swift specific version :
        - ./swift-saio.sh -r VERSION_HERE


    PLEASE NOTE: 
        * If -r flag is not specified it will default to 1.4.1 version 

        * SWIFT_DEPLOYMENT variable within the .cfg file default to package installation. 
          Packages are only built for Ubuntu 10.04 LTS (lucid)

        * If you would like to install from source please change SWIFT_DEPLOYMENT to "source" instead of "package"

        * The -6 flag will setup a full IPv6 swift SAIO setup. It will only work if the machine/VM uses 
          IPv6 global or ULA addresses. LocalLink addresses are not acceptable. 

        * This should be run on a clean/new virtual machine OR plain machine (not used for anything else) 
        


ChangeLog
swift-saio.sh (2012-03-09, ver=master)

    * Adding some new middleware options to the proxy and object configs
    * Also updating default version to 1.4.6

swift-saio.sh (2012-02-09, ver=master)

    * Adding object-expiring config 
    * Also updating default version to 1.4.5

swift-saio.sh (2011-10-28, ver=master)

    * Fixing a few more things and also testing it on 11.10 by doing a source installation

swift-saio.sh (2011-10-27, ver=master)

    * Added python-slogging support for 1.4.2 and above 
    * Also added the cronjobs for all the python-slogging needs
    * More information on : https://github.com/btorch/slogging/blob/master/doc/source/overview_stats.rst  
    * Removing all Launchpad PPA stuff and using just crashsite from now on

swift-saio.sh (2011-08-02, ver=master)

    * Added swift stats logging for 1.4.1 and above 
    * No cronjobs for the stats logging yet, users should run it manually for testing
    * Setting up syslog-ng to also collect stats logging informartion 

swift-saio.sh (2011-07-31, ver=master)

    * Added dispersion reporting configuration for 1.4.1 and above 

swift-saio.sh (2011-07-18, ver=master)

    * Added syslog-ng configuration that includes separation of swift logging
    * Also modified the log_facility for each service in order for the above to work
    * Fixed some minor typos, identation and some other cosmetic things

swift-saio.sh (2011-07-13, ver=master)

    * Added support for SSL for the proxy 
    * You can now specify in the swift-saio.cfg config file if you want SSL under the proxy section  
    * You can also change openssl certificate values within the swift-saio.sh/saio-openssl.cnf config




swift-saio.sh
#!/bin/bash

# Author: Marcelo Martins
# Date: 2011-05-01
# Version: 1.00
#
# Info:
#       Script for setting up an openstack-swift All-In-One environment  
#       Currently the script is setup to install swift from source
#
#

# CHECK USER ID (MUST BE ROOT)
if [ "$(id -u)" != "0" ]; then
   echo "This script must be run as root"
   exit 1
fi


##############################
# INITIALIZING SOME VARIABLE
##############################
PYTHON_VER="2.6"
BIND_IP="0.0.0.0"
IPV6_SUPPORT="false"
CURDIR=`pwd`
CFG_FILE="$CURDIR/swift-saio.cfg"
MODULES="$CURDIR/modules"
SYSLOGNG="$CURDIR/etc/syslog-ng"

# SOURCE CONFIGURATION FILE
if [ ! -e $CFG_FILE ]; then 
    printf "\n\t No configuration file found (\033[1;31;40m mmissing: %s \033[0m)\n" "$CFG_FILE" 
    exit 1
else
    #printf "\n\t - Sourcing configuration file "
    source $CFG_FILE
fi

if [[ -z $VERSION ]]; then 
    VERSION="$SWIFT_VER"
fi

TEMPLATE_VER=`echo ${VERSION:0:3}`".x"
TEMPLATES="./etc/swift-$TEMPLATE_VER"
PATCHES="$CURDIR/patches/$VERSION"

# ARGUMENTS 
NUMBER_OF_ARGS=$#
ARG_SARRAY=("$@")
ARGS=$@



usage_display (){
cat << USAGE

Syntax
    sudo swift-saio.sh [-r swift_version]  
    -r  The swift version to be installed 
    -6  Setup using ipv6 addresses (only for 1.3 and above)
    -h  For this usage screen  

USAGE
exit 1
}


while getopts "r:6h" opts
do 
    case $opts in 
        r) 
            VERSION="${OPTARG}"
            VER_REGEX="^1\.[2-9]\.*([0-9])*$"
            if [[ ! "$VERSION" =~ $VER_REGEX ]]; then 
                printf "\t Sorry, only version 1.2, 1.3 and 1.4 currently supported \n"
                exit 1 
            fi
            ;;
        6)
            VER_REGEX="^1\.[3-9]\.*([0-9])*$"
            if [[ ! "$VERSION" =~ $VER_REGEX ]]; then
                printf "\t Sorry, IPv6 only supported on version 1.3 and above \n"
                exit 1 
            else
                IPV6_SUPPORT="true"
                BIND_IP="::"
            fi
            ;;
        h)
            usage_display
            ;;
        *) 
            usage
            ;;
    esac     
done



##################
# FUNCTIONS 
##################
# Progress bar 
spinner (){
    PROC=$1
    echo -n '[' 
    while [ -d /proc/$PROC ];do
        echo -n '#' ; sleep 0.25
        echo -n '#' ; sleep 0.25
        echo -n '#' ; sleep 0.25
        echo -n '#' ; sleep 0.25
    done
    echo -n ' - 100%]' 
    return 0
}

main_banner (){

    SYSTEM=`cat /etc/issue | head -n1`
    ARCH=`uname -m`
    IP6=`lsmod |grep ipv6 >/dev/null ; echo $?`
    if [ $IP6 -eq 0 ]; then 
        IP6="yes"
    else
        IP6="no"
    fi

    printf "\n\t #######################################################"
    printf "\n\t #           WELCOME TO SWIFT-SAIO.SH ver1.0           #" 
    printf "\n\t #######################################################"
    printf "\n\t #   Your System : %s " "$SYSTEM"
    printf "\n\t #   Your Arch : %s " "$ARCH"
    printf "\n\t #   IPv6 Ready  : %s " "$IP6"
    printf "\n\t #  " 
    printf "\n\t #   Swift Install : %s " "$VERSION"
    printf "\n\t # "
    printf "\n\t # \n"
}   


check_version () {
    if [[ "$VERSION" =~ "^1\.2" ]]; then
        if [[ $SWIFT_DEPLOYMENT = "package" ]]; then 
            printf "\n\t - Sorry only Swift Version 1.4.1 allows package installation \n"
            exit 1
        fi
    elif [[ "$VERSION" =~ "^1\.3" ]]; then
        if [[ $SWIFT_DEPLOYMENT = "package" ]]; then
            printf "\n\t - Sorry only Swift Version 1.4.1 allows package installation \n"
            exit 1
        fi        
    fi
}


################## 
#      MAIN
##################

# Call Intro Banner
main_banner
check_version


####################################
#  CHECK IPV6   
####################################
if [ "$IPV6_SUPPORT" = "true" ]; then 
    source $MODULES/ipv6_config.sh
    ipv6_banner 
    ipv6_calc_install "ubuntu"
    ipv6_check
    sleep 2
fi


####################################
#  NON-PYTHON DEPENDENCIES INSTALL  
####################################
source $MODULES/non_python_deps_install.sh
install_non_python_deps "ubuntu"


####################################
#  PYTHON DEPENDENCIES INSTALL  
####################################
if [[ $SWIFT_DEPLOYMENT = "source" ]]; then 
    source $MODULES/python_deps_install.sh
    install_python_deps 
fi


########################
# SWIFT INSTALL 
########################
if [[ $SWIFT_DEPLOYMENT = "package" ]]; then 
    source $MODULES/swift_repo_install.sh
    swift_repo_install
elif [[ $SWIFT_DEPLOYMENT = "source" ]]; then     
    #source $MODULES/swift_source_install.sh
    #swift_source_install
    source $MODULES/swift_source_install_github.sh
    swift_source_install
fi     



########################
# SWAUTH INSTALL 
########################
if [[ $SWIFT_DEPLOYMENT = "package" ]]; then
    source $MODULES/swauth_deb_install.sh
    swauth_deb_install
elif [[ $SWIFT_DEPLOYMENT = "source" ]]; then
    source $MODULES/swauth_source_install.sh
    swauth_source_install
fi 


#########################
# STORAGE DEVICES SETUP 
#########################
source $MODULES/loopdev.sh
setup_loopdev $MOUNT_LOCATION $DD_FILE $DD_BS_VALUE $DD_SEEK_VALUE



###################################
# SWIFT CONFIGURATION & RING SETUP
###################################

# Configuration Path setup  
mkdir -p $SWIFT_CONF


printf "\n\t - Starting Swift $VERSION configuration setup "

if [ ! -d $TEMPLATES ]; then 
    printf "\t\t Unabled to locate template's directory \n"
    printf "\t\t Are you running the script from within it's source home ? \n"
    printf "\t\t Your current directory : `pwd` \n"
    printf "\t\t Quiting swift-saio setup \n"
    exit 1 
fi


printf "\n\t\t Setting up /etc/swift/swift.conf"
# All nodes need this file
cp $TEMPLATES/swift.conf.tmpl  /etc/swift/swift.conf
sed -i "s/HASH_PATH_SUFFIX/$HASH_PATH_SUFFIX/" /etc/swift/swift.conf


## Common to object/container/account services
printf "\n\t\t Setting up /etc/swift/drive-audit.conf"
cp $TEMPLATES/drive-audit.conf.tmpl  /etc/swift/drive-audit.conf
sed -i "s#MOUNTPOINT#$MOUNT_LOCATION#" /etc/swift/drive-audit.conf


## Rsync setup
printf "\n\t\t Setting up rsync configs \n\n"
sed -i 's/RSYNC_ENABLE=false/RSYNC_ENABLE=true/' /etc/default/rsync
cp $TEMPLATES/rsyncd.conf.tmpl   /etc/rsyncd.conf
sed -i "s#MOUNTPOINT#$MOUNT_LOCATION#g"  /etc/rsyncd.conf


###############
# Object-server
###############
source $MODULES/object.sh
setup_object 


##################
# Container-server 
##################
source $MODULES/container.sh
setup_container


#################
# Account-server
#################
source $MODULES/account.sh
setup_account


###############
# Proxy-server 
###############
source $MODULES/proxy.sh
setup_proxy


###############
# SWIFT FINALE 
###############
source $MODULES/final_steps.sh
final_steps

source $MODULES/swift_syslog-ng_setup.sh
swift_syslog_ng_setup


##################
# START SERVICES 
##################
source $MODULES/start_services.sh
start_services


###############################
# SETUP DISPERSION TOOL 
###############################
source $MODULES/swift_dispersion_reporting.sh
dispersion_setup


###############################
# SETUP SWAUTH & ADMIN ACCOUNT
###############################
source $MODULES/swauth_setup.sh
swauth_setup
sleep 5


###############################
# SETUP SLOGGING 
###############################
source $MODULES/swift_stats_logging_setup.sh
slogging_setup


###############################
# SWAUTH ACCOUNT INFO/URL
###############################
swauth_info


exit 0 
